cmd/site-replication.go

```diff
@@ -5516,3 +5872,79 @@ func (c *SiteReplicationSys) getSiteMetrics(ctx context.Context) (madmin.SRMetri
 	sm.Uptime = UTCNow().Unix() - globalBootTime.Unix()
 	return sm, nil
 }
+
+// mergeWithCurrentLCConfig - merges the given ilm expiry configuration with existing for the current site and returns
+func mergeWithCurrentLCConfig(ctx context.Context, bucket string, expLCCfg *string, updatedAt time.Time) ([]byte, error) {
+	// Get bucket config from current site
+	meta, e := globalBucketMetadataSys.GetConfigFromDisk(ctx, bucket)
+	if e != nil && !errors.Is(e, errConfigNotFound) {
+		return []byte{}, e
+	}
+	rMap := make(map[string]lifecycle.Rule)
+	var xmlName xml.Name
+	if len(meta.LifecycleConfigXML) > 0 {
+		var lcCfg lifecycle.Lifecycle
+		if err := xml.Unmarshal(meta.LifecycleConfigXML, &lcCfg); err != nil {
+			return []byte{}, err
+		}
+		for _, rl := range lcCfg.Rules {
+			rMap[rl.ID] = rl
+		}
+		xmlName = meta.lifecycleConfig.XMLName
+	}
+
+	// get latest expiry rules
+	newRMap := make(map[string]lifecycle.Rule)
+	if expLCCfg != nil {
+		var cfg lifecycle.Lifecycle
+		expLcCfgData, err := base64.StdEncoding.DecodeString(*expLCCfg)
+		if err != nil {
+			return []byte{}, err
+		}
+		if err := xml.Unmarshal(expLcCfgData, &cfg); err != nil {
+			return []byte{}, err
+		}
+		for _, rl := range cfg.Rules {
+			newRMap[rl.ID] = rl
+		}
+		xmlName = cfg.XMLName
+	}
+
+	// check if current expiry rules are there in new one. if not remove as they may
+	// have been removed from latest updated one
+	for id, rl := range rMap {
+		if !rl.Expiration.IsNull() || !rl.NoncurrentVersionExpiration.IsNull() {
+			if _, ok := newRMap[id]; !ok {
+				delete(rMap, id)
+			}
+		}
+	}
+
+	// append now
+	for id, rl := range newRMap {
+		rMap[id] = rl
+	}
+
+	var rules []lifecycle.Rule
+	for _, rule := range rMap {
+		rules = append(rules, rule)
+	}
+
+	// no rules, return
+	if len(rules) == 0 {
+		return []byte{}, nil
+	}
+
+	// get final list for write
+	finalLcCfg := lifecycle.Lifecycle{
+		XMLName:         xmlName,
+		Rules:           rules,
+		ExpiryUpdatedAt: &updatedAt,
+	}
```

Comment:
validate the finalLcCfg before matshalling

Reply:
Ack